src/lib/auth/oauth-client.ts

```diff
@@ -0,0 +1,170 @@
+import http, { IncomingMessage, ServerResponse } from "http";
+const { randomBytes, createHash } = require("node:crypto");
+import url from "url";
+import net from "net";
+
+const accountURL = process.env.FAUNA_ACCOUNT_URL ?? "https://account.fauna.com";
+
+// Default to prod client id and secret
+const clientId = process.env.FAUNA_CLIENT_ID ?? "Aq4_G0mOtm_F1fK3PuzE0k-i9F0";
+// Native public clients are not confidential. The client secret is not used beyond
+//   client identification. https://datatracker.ietf.org/doc/html/rfc8252#section-8.5
+const clientSecret =
+  process.env.FAUNA_CLIENT_SECRET ??
+  "2W9eZYlyN5XwnpvaP3AwOfclrtAjTXncH6k-bdFq1ZV0hZMFPzRIfg";
+const REDIRECT_URI = `http://127.0.0.1`;
+
+class OAuthClient {
+  public server: http.Server;
+  public port: number;
+  private code_verifier: string;
+  private code_challenge: string;
+  private auth_code: string;
+  public state: string;
+
+  constructor() {
+    this.server = http.createServer(this._handleRequest.bind(this));
+    this.code_verifier = Buffer.from(randomBytes(20)).toString("base64url");
+    this.code_challenge = createHash("sha256")
+      .update(this.code_verifier)
+      .digest("base64url");
+    this.port = 0;
+    this.auth_code = "";
+    this.state = this._generateCSRFToken();
+  }
+
+  public getRequestUrl() {
+    const params = {
+      client_id: clientId,
+      redirect_uri: `${REDIRECT_URI}:${this.port}`,
+      code_challenge: this.code_challenge,
+      code_challenge_method: "S256",
+      response_type: "code",
+      scope: "create_session",
+      state: this.state,
+    };
+    return `${accountURL}/api/v1/oauth/authorize?${new URLSearchParams(
+      params
+    )}`;
+  }
+
+  public getToken() {
+    const now = new Date();
+    // Short expiry for access token as it's only used to create a session
+    now.setUTCMinutes(now.getUTCMinutes() + 5);
```

Comment:
@cleve-fauna should we enforce this in the backend where frontdoor looks at the clientID and overrides any ttl param with our own

Reply:
probably since there's nothing stopping ppl from having a token that doesn't expire and continuing to create sessions